### PR TITLE
Conserve memory for large data work and migrate to Parsl

### PR DESCRIPTION
**Pull request by d33bs**
# Description

This PR seeks to further address #43 by conserving the use of hardware resources through the use of `chunk_size` and parallel operations. Apologies in advance for the large size of the changes, many of these became necessary in working through the challenges. My hope is that these changes make CytoTable perform in a more stable and performant way moving forward.

I'll outline notable aspects below:

- `chunk_size` is now used for initial data export and joining operations to parquet format (thank you @gwaybio for the helpful example from #43). Moving to "chunked" work in this way influenced a desire to address parallel operations (see below).
- Parsl is now used as the underlying workflow engine to assist with workflow orchestration. `parsl_config` was added as a way to configure Parsl for use by CytoTable.
  - Notes: As development for this PR was in progress, using Prefect within a local environment (both with or without the Prefect API agent running) was often unstable, sometimes resulting in exceptions or cancelled workflows for larger datasets. As a result, I migrated existing work to use Parsl and reorganized execution of functions to abide by Parsl's constraints. Parsl was selected due to flexible execution environment options which may align well with the audience using this work and similarity to Prefect's pythonic implementation (requiring fewer changes than other workflow engines).
- Preset configurations may be intermixed with specific parameter overrides through `convert` (many thanks @jenna-tomkinson for help in understanding potential needs here).
- Data ingest was refactored to utilize DuckDB for both CSV and SQLite data sources in order to take advantage of DuckDB's performance offerings and unify the process of reading as well as writing data.
- Testing was updated to reflect new workflows and internal data structures.

## What is the nature of your change?

- [x] Bug fix (fixes an issue).
- [x] Enhancement (adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [x] This change requires a documentation update.

# Checklist

Please ensure that all boxes are checked before indicating that a pull request is ready for review.

- [x] I have read the [CONTRIBUTING.md](CONTRIBUTING.md) guidelines.
- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have deleted all non-relevant text in this pull request template.

### CODE REVIEW

File: tests/test_convert.py [hunk 1]
```diff
@@ -775,18 +747,35 @@ def test_convert_cellprofiler_csv(
...
+    Tests convert with Parsl HighThroughputExecutor
 
-    Note: dedicated test for prefect-dask runner
+    See the following for more details.
+    https://parsl.readthedocs.io/en/stable/stubs/parsl.executors.HighThroughputExecutor.html#parsl.executors.HighThroughputExecutor
     """
 
+    local_htex = Config(
+        executors=[
+            HighThroughputExecutor(
```
Comment (gwaybio) [inline tests/test_convert.py:765]: nice to see this test! 🎉

**Review by gwaybio — COMMENTED** [review]
Very minor comments - congrats on this PR!

Let's let @falquaddoomi give this a more technical look. Thanks!

File: .gitignore [hunk 2]
```diff
@@ -139,3 +139,6 @@ cue.mod/dagger*
 
 # dask
 dask-worker-space
+
+#parsl
```
[the remark this answers is not in the document]
Comment (d33bs) [reply .gitignore:143]: Thanks - yes, that's correct a comment to indicate what the ignore below this line is for. I'll update this line as per your suggestion.

File: cytotable/convert.py [hunk 3]
```diff
@@ -2,89 +2,175 @@
...
+    # return the filepath for the chunked output file
+    return result_filepath
+
+
+@python_app
+def _prepend_column_name(
+    table_path: str,
     source_group_name: str,
-    identifying_columns: Union[List[str], Tuple[str, ...]],
+    identifying_columns: List[str],
```
[the remark this answers is not in the document]
Comment (d33bs) [reply cytotable/convert.py:164]: Thank you! I'll update in the docs!

File: cytotable/convert.py [hunk 4]
```diff
@@ -253,101 +353,57 @@ def _concat_source_group(
...
-        destination_path = pathlib.Path(
-            (
-                f"{dest_path}/{str(source['source_path'].parent.name)}"
-                f".{str(source['source_path'].stem)}.parquet"
-            )
-        )
-
-    # write the table to destination path output
-    parquet.write_table(table=source["table"], where=destination_path)
+            # pathlib.Path(pathlib.Path(source["table"][0]).parent).rmdir()
```
[the remark this answers is not in the document]
Comment (d33bs) [reply cytotable/convert.py:398]: Great catch, thanks! I'll remove this.

File: tests/test_convert.py (hunk 1 above)
Comment (d33bs) [reply tests/test_convert.py:765]: Cheers - exciting possibilities here!

File: cytotable/convert.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (d33bs) [reply cytotable/convert.py]: On a second look at this I chose to re-include this line with more elegant handling for scenarios where directories may not be safe to remove. The focus here is to remove subdirectories within a compartment or metadata directory. When concatenating multiple subdirectories of data from the same compartment or metadata tables we can end up with one to many empty subdirectories after concatenation (the concatted result won't be nested so deeply).

For ex, I'd expect this to be the output of a "concat but not join" operation:
```text
destination_dir
├── cells
│   └── cells.parquet
├── cytoplasm
│   └── cytoplasm.parquet
├── image
│   └── image.parquet
└── nuclei
    └── nuclei.parquet
```

Without the added line to remove "parent" table directories we might end up with something like this:
```text
destination_dir
├── cells
│   └── subdir1
│   └── cells.parquet
├── cytoplasm
│   └── subdir1
│   └── cytoplasm.parquet
├── image
│   └── subdir1
│   └── image.parquet
└── nuclei
    └── subdir1
    └── nuclei.parquet
```

The extra directories are maybe mostly harmless but worth removing to keep the results clean.

**Review by falquaddoomi — APPROVED** [review]
Awesome work! It'll be exciting to see how Parsl's possibilities play out as we deal with larger datasets in possibly more distributed environments. Kudos for pivoting from Prefect so quickly, too.

**Review by falquaddoomi — APPROVED** [review]
Awesome work! It'll be exciting to see how Parsl's possibilities play out as we deal with larger datasets in possibly more distributed environments. Kudos for pivoting from Prefect so quickly, too.

**Comment by d33bs** [comment]
Just a minor note for clarity: I updated the testing suite to run the HighThroughputExecutor test (instead of skipping) and also the tests themselves to run in the context of how this project will be used.

**Pull request merged**